_posts/2015-08-07-symbolic-operators.md

```diff
@@ -0,0 +1,183 @@
+19 |+| 20 // produces 39
+```
+
+How does this work? We know that the `Int` type does not have a `|+|` method.
+Experienced Scala developers will suspect that implicits play a role here,
+but what are the details?
+
+In detail
+---------
+
+Let's walk through how the expression `19 |+| 20` is compiled.
+
+First, a `|+|` method is needed on `Int`. Since
+[`Int`](https://github.com/scala/scala/blob/v2.11.7/src/library/scala/Int.scala)
+does not provide one, the compiler searches for an implicit conversion to a
+type that *does* have a `|+|` method.
+
+Due to our import, it will find the
+[`semigroupSyntax[A]`](https://github.com/non/cats/blob/82dbf4076572dfbb6e29dd49875f5e5d929f80be/core/src/main/scala/cats/syntax/semigroup.scala#L8)
+method, which returns a type that has a `|+|` method (specifically
+[`SemigroupOps[A]`](https://github.com/non/cats/blob/82dbf4076572dfbb6e29dd49875f5e5d929f80be/core/src/main/scala/cats/syntax/semigroup.scala#L12)).
+However, `semigroupSyntax` requires an implicit `Semigroup[A]` value to be in scope.
+Do we have a `Semigroup[Int]` in scope?
+
+Yes we do. Our import also provides an implicit value [`intGroup`](https://github.com/non/cats/blob/82dbf4076572dfbb6e29dd49875f5e5d929f80be/std/src/main/scala/cats/std/anyval.scala#L23)
+of type `AdditiveCommutativeGroup[Int]`. Leaving aside what *additive*, *commutative*,
+and *group* mean here, this is a subtype of `Semigroup[Int]`, so it matches.
+
+Let's continue with our current example. At this point we have gone form:
+
+```scala
+19 |+| 20 // produces 39
+```
+
+to:
+
+```scala
+semigroupSyntax[Int](19)(intGroup) |+| 20
+```
+
+But we aren't out of the woods yet! We still need to see how this expression
+is evaluated.
+
+Of macros and machinists
+------------------------
+
+Looking at how the `|+|` method is
+[implemented](https://github.com/non/cats/blob/82dbf4076572dfbb6e29dd49875f5e5d929f80be/core/src/main/scala/cats/syntax/semigroup.scala#L13)
+reveals the cryptic `macro Ops.binop[A, A]`. What is this?
+
+Following the rabbit hole farther, we come to
+[`cats.macros.Ops`](https://github.com/non/cats/blob/82dbf4076572dfbb6e29dd49875f5e5d929f80be/macros/src/main/scala/cats/macros/Ops.scala#L6)
+which provides the macro implementation that `|+|` is using. Aside from a
+[suggestively named](https://github.com/non/cats/blob/82dbf4076572dfbb6e29dd49875f5e5d929f80be/macros/src/main/scala/cats/macros/Ops.scala#L18)
+item in the `operatorNames` map, we don't have any clues what is going on.
+
+The [machinist](https://github.com/typelevel/machinist) project was created
+to optimize exactly this kind of implicit syntax problem. What will happen here
+is that `operatorNames` describes how to rewrite expressions using type
+classes. Long-story short, we will transform:
+
+```scala
+semigroupSyntax[Int](19)(intGroup) |+| 20
+```
+
+into:
+
+```scala
+intGroup.combine(19, 20)
+```
+
+The suggestive map item tells us that we should rewrite the `|+|` operator
```

Comment:
Maybe, "The *aforementioned* suggestive map item", or some other connective tissue to tie it to the `operatorNames` reference above?